### PR TITLE
Taulink taranis telemetry

### PR DESCRIPTION
**Pull request by elafargue**
This implements support for Taranis Telemetry display from a TauLink (all code from Peabody124).

### CODE REVIEW

**Comment by DTFUHF** [comment]
Awesome addition.

I think the s.port telem TX pin inversion should be programmable. This would help out the people who have done the 9X telemetry conversion. It'd also allow the use of the usart for regular non-inverted applications.

An example of this is here in my branch https://github.com/DTFUHF/TauLabs/commit/29e1d9e316d29fae842da9b3d8ae5274d9b68ce2 that also adds a struct to allow general usart inversion in other cases. This uses PA15 on pipxtreme to drive what's assumed to be an XOR. PA15 is currently unused and doesn't use any future peripherals (i2c/SPI/CAN etc)

File: flight/Modules/UAVOTaranis/UAVOTaranis.c [hunk 1]
```diff
@@ -0,0 +1,898 @@
...
+		*value = (rfm22bStatus.LinkQuality < local_link_quality) ? rfm22bStatus.LinkQuality : local_link_quality;
+		
+		// Rescale to values that match Taranis
+		if (*value < 64) {
+			*value = 0;
+		} else {
+			*value = (*value - 64) * 100 / 64;
+		}
+	} else {
+		*value = 0;
```
Comment (guilhermito) [inline flight/Modules/UAVOTaranis/UAVOTaranis.c]: shouldn't false be returned here?

File: flight/Modules/UAVOTaranis/UAVOTaranis.c [hunk 2]
```diff
@@ -0,0 +1,898 @@
...
+		return false;
+
+	if (test_presence_only)
+		return true;
+
+	AttitudeActualData attitude;
+	AttitudeActualGet(&attitude);
+	float hdg = (attitude.Yaw >= 0) ? attitude.Yaw : (attitude.Yaw + 360.0f);
+	*value = (uint32_t)(hdg * 100.0f);
+	*/
```
Comment (guilhermito) [inline flight/Modules/UAVOTaranis/UAVOTaranis.c]: remove commented out code

File: flight/Modules/UAVOTaranis/UAVOTaranis.c [hunk 3]
```diff
@@ -0,0 +1,898 @@
...
+			t2 = 400;
+		else
+			t2 = 300;
+		break;
+	}
+	if (frsky->gps_position.Satellites > 0)
+		t2 += frsky->gps_position.Satellites;
+
+	*value = (uint32_t)t2;
+	*/
```
Comment (guilhermito) [inline flight/Modules/UAVOTaranis/UAVOTaranis.c]: remove commented code

File: flight/Modules/UAVOTaranis/UAVOTaranis.c [hunk 4]
```diff
@@ -0,0 +1,898 @@
...
+		AccelszGet(&acc);
+		break;
+	}
+
+	acc /= GRAVITY;
+	acc *= 100.0f;
+
+	int32_t frsky_acc = (int32_t) acc;
+	*value = (uint32_t) frsky_acc;
+	*/
```
Comment (guilhermito) [inline flight/Modules/UAVOTaranis/UAVOTaranis.c]: remove commented code

File: flight/Modules/UAVOTaranis/UAVOTaranis.c [hunk 5]
```diff
@@ -0,0 +1,898 @@
...
+		if (coord >= 0)
+			frsky_coord = 2 << 30;
+		else
+			frsky_coord = 3 << 30;
+	}
+	coord = abs(coord);
+	frsky_coord |= (((uint64_t)coord * 6ull) / 100);
+
+	*value = frsky_coord;
+	*/
```
Comment (guilhermito) [inline flight/Modules/UAVOTaranis/UAVOTaranis.c]: remove commented code

File: flight/Modules/UAVOTaranis/UAVOTaranis.c [hunk 6]
```diff
@@ -0,0 +1,898 @@
...
+	/*if (GPSPositionHandle() == NULL)
+		return false;
+	if (frsky->gps_position.Status != GPSPOSITION_STATUS_FIX3D
+			&& frsky->gps_position.Status != GPSPOSITION_STATUS_DIFF3D)
+		return false;
+	if (test_presence_only)
+		return true;
+
+	int32_t frsky_gps_alt = (int32_t)(frsky->gps_position.Altitude * 100.0f);
+	*value = (uint32_t)frsky_gps_alt;*/
```
Comment (guilhermito) [inline flight/Modules/UAVOTaranis/UAVOTaranis.c]: remove commented code

File: flight/Modules/UAVOTaranis/UAVOTaranis.c [hunk 7]
```diff
@@ -0,0 +1,898 @@
...
+ * \return -1 if initialization failed
+ * \return 0 on success
+ */
+static int32_t uavoTaranisInitialize(void)
+{
+	uint32_t sport_com = PIOS_COM_FRSKY_SPORT;
+
+	if (sport_com) {
+
+
```
Comment (guilhermito) [inline flight/Modules/UAVOTaranis/UAVOTaranis.c]: extra lines

File: flight/Modules/UAVOTaranis/UAVOTaranis.c [hunk 8]
```diff
@@ -0,0 +1,898 @@
...
+			PIOS_Thread_Sleep(25);
+			frsky_encode_rssi(&value, false, 0);
+			frsky_send_frame(FRSKY_RSSI_ID, value);
+
+			PIOS_Thread_Sleep(25);
+			frsky_encode_swr(&value, false, 0);
+			frsky_send_frame(FRSKY_SWR_ID, value);
+
+		}
```
Comment (guilhermito) [inline flight/Modules/UAVOTaranis/UAVOTaranis.c]: Debug code above?

File: flight/Modules/UAVOTaranis/UAVOTaranis.c [hunk 9]
```diff
@@ -0,0 +1,898 @@
...
+			// for some reason, only first four messages are sent.
+			for (uint32_t i = 0; i < sizeof(frsky_value_items) / sizeof(frsky_value_items[0]); i++) {
+				frsky->scheduled_item = i;
+				frsky_send_scheduled_item();
+				PIOS_Thread_Sleep(5);
+			}
+
+		}
+
+		if (false) {
```
Comment (guilhermito) [inline flight/Modules/UAVOTaranis/UAVOTaranis.c]: again?

File: flight/targets/pipxtreme/board-info/board-info.mk [hunk 10]
```diff
@@ -1,5 +1,5 @@
 BOARD_TYPE          := 0x03
-BOARD_REVISION      := 0x01
+BOARD_REVISION      := 0x02
```
Comment (guilhermito) [inline flight/targets/pipxtreme/board-info/board-info.mk]: I'm confused, why is this here? are you using the pipx board file for the taulink?

File: flight/targets/pipxtreme/board-info/board_hw_defs.c [hunk 11]
```diff
@@ -478,6 +518,50 @@ static const struct pios_usart_cfg pios_usart_serial_cfg =
 	},
 };
 
+static const struct pios_usart_cfg pios_usart_bluetooth_cfg =
+{
+	.regs = USART2,
+	.init =
+	{
+		.USART_BaudRate = 9600,
```
Comment (guilhermito) [inline flight/targets/pipxtreme/board-info/board_hw_defs.c]: Any reason the limit this to 9600?

File: flight/targets/pipxtreme/board-info/pios_board.h [hunk 12]
```diff
@@ -169,20 +172,26 @@ extern uintptr_t pios_com_telem_usb_id;
...
 #define PIOS_COM_TELEM_UART_FLEXI  (pios_com_telem_uart_flexi_id)
 #define PIOS_COM_TELEM_UART_TELEM  (pios_com_telem_uart_main_id)
+#define PIOS_COM_BRIDGE            (pios_com_bridge_id)
+#define PIOS_COM_VCP               (pios_com_vcp_id)
 #define PIOS_COM_TELEMETRY         (pios_com_telemetry_id)
 #define PIOS_COM_RFM22B            (pios_com_rfm22b_id)
 #define PIOS_COM_RADIO             (pios_com_radio_id)
 #define PIOS_PPM_RECEIVER          (pios_ppm_rcvr_id)
 #define PIOS_COM_DEBUG             (pios_com_debug_id)
+#define PIOS_COM_FRSKY_SPORT            (pios_com_frsky_sport_id)
```
Comment (guilhermito) [inline flight/targets/pipxtreme/board-info/pios_board.h]: extra tab?

File: flight/targets/pipxtreme/fw/pios_board.c [hunk 13]
```diff
@@ -245,6 +308,98 @@ void PIOS_Board_Init(void) {
...
+            PIOS_COM_ChangeBaud(pios_com_telem_uart_bluetooth_id, comBaud);
+
+            // Note this doesn't actually send until RTOS is running
+            PIOS_COM_SendString(pios_com_telem_uart_bluetooth_id,"AT");
+            PIOS_COM_SendString(pios_com_telem_uart_bluetooth_id,"AT+NAMETauLink");
+            PIOS_COM_SendString(pios_com_telem_uart_bluetooth_id,"AT+PIN:000000");
+            PIOS_COM_SendString(pios_com_telem_uart_bluetooth_id,"AT+MODE1");
+            PIOS_COM_SendString(pios_com_telem_uart_bluetooth_id,"AT+SHOW1");
+            PIOS_COM_SendString(pios_com_telem_uart_bluetooth_id,"AT+RESET");
+            PIOS_COM_SendString(pios_com_telem_uart_bluetooth_id,"AT+BAUD4"); // 115200
```
Comment (guilhermito) [inline flight/targets/pipxtreme/fw/pios_board.c]: "AT+BAUD4" is not 115200bps

**Comment by elafargue** [comment]
@guilhermito -> I will let @peabody124 comment, it's all his code :)

**Comment by mlyle** [comment]
OK, I read this.  It's all duplicated/copied code from elsewhere.  Either stuff needs to get factored into a library or the UAVOFrSKYSPortBridge code needs to be taught to work in this circumstance too.

Copy and paste must die.

**Comment by elafargue** [comment]
Not sure what you meant, mlyle? There is another way to do Taranis telemetry on the current codebase? I'm a bit out of it at the moment, so there might very well be and I'm not aware of it...

**Comment by mlyle** [comment]
@elafargue Hey-- just to clarify, I closed this for two reasons.

First, @peabody124 says it's not to merge because it depends on TLM hardware which is not (yet) open.  Since this is still in limbo the PR can't merge for some time.

Second, I think it should not merge because UAVOTaranis is a copy of UAVOFrSkySPortBridge with some blocks commented out and a couple of other changes.  This is bad because UAVOFrSkySensorHubBridge has since changed significantly since the point at which UAVOTaranis was copied, and because it bloats the code and makes it unmaintainable.

It's a little bit more initial work to teach the UAVOFrSkySPortBridge to work in both cases, but it means that anyone changing FrSky stuff in the future doesn't need to change it in two (or three-- because there's some degree of duplication with UAVOFrSkySensorHubBridge even).

I've been working a lot on removing code duplication and it's very difficult after the fact to figure out why differences are in nearly-identical code-- are they deliberate or is it just the result of drift, etc.  I really want us to avoid adding more copy-and-paste to the codebase.
